nemo/collections/nlp/models/nlp_model.py

```diff
@@ -410,3 +522,160 @@ def save_checkpoint(self, filepath, weights_only: bool):
                     )
                     atomic_save(checkpoint, filepath)
         return None
+
+
+class NLPDDPPlugin(DDPPlugin):
+    """ DDP plugin for Pytorch Lightning. Needed to customize DDP for model parallel models.
+    """
+
+    distributed_backend = "ddp"
+
+    def __init__(
+        self,
+        parallel_devices: Optional[List[torch.device]] = None,
+        num_nodes: int = 1,
+        cluster_environment: ClusterEnvironment = None,
+        sync_batchnorm: bool = False,
+        **kwargs: Union[Any, Dict[str, Any]],
+    ) -> None:
+        super().__init__(parallel_devices, num_nodes, cluster_environment, sync_batchnorm, **kwargs)
+
+    def init_ddp_connection(self, global_rank: int, world_size: int) -> None:
+        # call PTL init ddp
+        super().init_ddp_connection(global_rank, world_size)
+
+        # init model parallel
+        app_state = AppState()
+
+        if app_state.model_parallel_size is not None:
+
+            if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+
+                if app_state.model_parallel_group is None:
+                    self.init_model_parallel(app_state.global_rank, app_state.world_size)
+
+    def start_training(self, trainer: 'Trainer') -> None:
+        """ PTL Hook that is called after DPP is initialized. """
+
+        if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+            app_state = AppState()
+            if app_state.model_parallel_size is not None:
+                # mpu grad clipping needs parameters to have the attribute model_parallel
+                parameters = self.lightning_module.parameters()
+                for p in parameters:
+                    if not hasattr(p, 'model_parallel'):
+                        p.model_parallel = False
+
+                # TODO: figure out how to override clip gradients again
```

Comment:
Is it fine to train Megatron without it? If not, maybe setup a warning here for the timebeing for megatron models.

Reply:
It's used in the megatron-lm gradient clipping code.